pkg/apis/kops/validation/openstack.go

```diff
@@ -0,0 +1,35 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/kops/pkg/apis/kops"
+)
+
+func openstackValidateCluster(c *kops.Cluster) (errList field.ErrorList) {
+	if c.Spec.CloudConfig.Openstack.Router != nil && c.Spec.CloudConfig.Openstack.Router.ExternalNetwork == nil {
```

Comment:
But public topology doesn't require an external network when there's no router?

Reply:
Public topology means using floating IPs. In order to use floating IPs you need both an external network and you need a router (which handles traffic between internal and external network). So public topology is not possible without one.

Reply:
But this validation code skips the topology check if `c.Spec.CloudConfig.Openstack.Router` is `nil`

Reply:
Ah. That is quite a bug. Fixed now. And added some tests as well.